### ISSUE TITLE
FP unnecessary-operation, #![feature(capture_disjoint_fields)]

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9896]
### Summary

.

### Lint Name

unnecessary-operation

### Reproducer

I tried this code:

```rust
#![allow(incomplete_features)]
#![feature(capture_disjoint_fields)]

fn foo<MsU>(handler: impl FnOnce() -> MsU + Clone + 'static) {
    Box::new(move |value| {
        (|_| handler.clone()())(value);
        None
    }) as Box<dyn Fn(i32) -> Option<i32>>;
}

fn main() {}

```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::unnecessary_operation`
```
    Checking clpy v0.1.0 (/tmp/clpy)
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
 --> src/main.rs:5:20
  |
5 |     Box::new(move |value| {
  |                    ^^^^^
  |
help: consider giving this closure parameter an explicit type
  |
5 |     Box::new(move |value: _| {
  |                         +++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: function `foo` is never used
 --> src/main.rs:4:4
  |
4 | fn foo<MsU>(handler: impl FnOnce() -> MsU + Clone + 'static) {
  |    ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: unnecessary operation
 --> src/main.rs:5:5
  |
5 | /     Box::new(move |value| {
6 | |         (|_| handler.clone()())(value);
7 | |         None
8 | |     }) as Box<dyn Fn(i32) -> Option<i32>>;
  | |__________________________________________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_operation
  = note: requested on the command line with `-W clippy::unnecessary-operation`
help: statement can be reduced to
  |
5 ~     Box::new(move |value| {
6 +         (|_| handler.clone()())(value);
7 +         None
8 +     });
  |

warning: unused boxed `std::ops::Fn` trait object that must be used
 --> src/main.rs:5:5
  |
5 | /     Box::new(move |value| {
6 | |         (|_| handler.clone()())(value);
7 | |         None
8 | |     }) as Box<dyn Fn(i32) -> Option<i32>>;
  | |_________________________________________^
  |
  = note: closures are lazy and do nothing unless called
  = note: `#[warn(unused_must_use)]` on by default

warning: `clpy` (bin "clpy") generated 3 warnings (run `cargo fix --bin "clpy"` to apply 1 suggestion)
warning: `clpy` (bin "clpy" test) generated 3 warnings (3 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 0.47s
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
